File: Arrays/Search & Sort/insertion-sorted-array.ts
// Insertion in a sorted array
// http://www.geeksforgeeks.org/search-insert-and-delete-in-a-sorted-array/

/**
 * Author: Arunkumar Srisailapathi
 * Email: sail.arunkumar@gmail.com
 * Web: arunkumars08.github.io
 */

/**
 * Input: [1, 2, 3, 5, 8] and insert 4
 */

class InsertionSortedArray {
    constructor(arr: Array<number>, insert: number) {
        const getPosition: number = new ModifiedBinarySearch().find(arr, insert, 0, arr.length - 1);
        const len: number = arr.length;

        if (getPosition + 1 === len) {
            if (arr[getPosition] <= insert) {
                arr.push(insert);
            } else {
                arr[getPosition + 1] = arr[getPosition];
                arr[getPosition] = insert;
            }
        } else {
            for (let i = len - 1; i >= getPosition; -- i) {
                arr[i + 1] = arr[i];
            }
            arr[getPosition + 1] = insert;
        }
        console.log(arr);
    }
}

class ModifiedBinarySearch {
    find(arr: Array<number>, insert: number, low: number, high: number): number {
        let mid: number = Math.floor((low + high) / 2);
        if (high === mid) return high;
        if (high < low) return low;

        if (insert < arr[mid]) {
            return this.find(arr, insert, low, mid - 1);
        } else {
            return this.find(arr, insert, mid + 1, high);
        }
    }
}

File: Arrays/Search & Sort/insertion-sorted-array.py
# Insertion in a sorted array
# http://www.geeksforgeeks.org/search-insert-and-delete-in-a-sorted-array/
# Author: Srivathsan
# Email: srivathsankr7@gmail.com

def find_position(l, h, arr, val):
    if h < l:
        return l
    mid = int(l + (h - l)/ 2)
    if h == mid:
        return h
    if arr[mid] == val:
        return mid
    elif arr[mid] > val:
        return find_position(l, mid-1, arr, val)
    else:
        return find_position(mid+1, h, arr, val)

def insert(arr, val):
    pos = find_position(0, len(arr)-1, arr, val)
    for i in range(pos, len(arr)):
        temp = arr[i]
        arr[i] = val
        val = temp
    arr.append(val)
    return arr


if __name__ == '__main__':
    arr =  [1, 2, 3, 5, 8] 
    newarr = insert(arr, 4)
    print(newarr)